backend/tests/v5/e2e/routes/sso/aad/aad.test.js

```diff
@@ -0,0 +1,180 @@
+			expect(redirectUri).toEqual(
+				expect.stringContaining('https://login.microsoftonline.com/common/oauth2/v2.0/authorize'),
+			);
+			expect(redirectUri).toEqual(
+				expect.stringContaining(`redirect_uri=${encodeURIComponent(authenticateRedirectUri)}`),
+			);
+			expect(redirectUri).toEqual(
+				expect.stringContaining(`state=${encodeURIComponent(JSON.stringify({ redirectUri: signupUri }))}`),
+			);
+		});
+	});
+};
+
+const testAuthenticatePost = () => {
+	describe('Sign Up Authenticate Post', () => {
+		test('should redirect the user to ', async () => {
+			const state = { redirectUri: generateRandomString() };
+			const res = await agent.get(`/v5/sso/aad/authenticate-post?state=${encodeURIComponent(JSON.stringify(state))}`)
+				.expect(templates.found.status);
+			expect(res.headers.location).toEqual(expect.stringMatching(state.redirectUri));
+		});
+	});
+};
+
+const signup = () => {
+	describe('Sign Up', () => {
+		const newUserData = {
+			username: generateRandomString(),
+			countryCode: 'GB',
+			company: generateRandomString(),
+			mailListAgreed: true,
+		};
+
+		test('should fail if the username already exists', async () => {
+			const res = await agent.post('/v5/sso/aad/signup').send({ ...newUserData, username: testUser.user })
+				.expect(templates.invalidArguments.status);
+			expect(res.body.code).toEqual(templates.invalidArguments.code);
+		});
+
+		test('should fail if there are missing body params', async () => {
+			const res = await agent.post('/v5/sso/aad/signup').send({ ...newUserData, username: undefined })
+				.expect(templates.invalidArguments.status);
+			expect(res.body.code).toEqual(templates.invalidArguments.code);
+		});
+
+		test('should validate signup data and redirect the user to Microsoft authentication page', async () => {
+			const res = await agent.post('/v5/sso/aad/signup').send(newUserData)
+				.expect(templates.found.status);
+
+			const redirectUri = res.headers.location;
+			expect(redirectUri).toEqual(
+				expect.stringContaining('https://login.microsoftonline.com/common/oauth2/v2.0/authorize'),
+			);
+			expect(redirectUri).toEqual(
+				expect.stringContaining(`redirect_uri=${encodeURIComponent(signupRedirectUri)}`),
+			);
+			expect(redirectUri).toEqual(
+				expect.stringContaining(`state=${encodeURIComponent(JSON.stringify({
+					username: newUserData.username,
+					countryCode: newUserData.countryCode,
+					company: newUserData.company,
+					mailListAgreed: newUserData.mailListAgreed,
+				}))}`),
+			);
+		});
+	});
+};
+
+const signupPost = () => {
+	describe('Sign Up', () => {
+		const newUserData = {
+			username: generateRandomString(),
+			countryCode: 'GB',
+			company: generateRandomString(),
+			mailListAgreed: true,
+		};
+
+		const newUserDataFromAad = {
+			mail: newUserEmail,
+			givenName: generateRandomString(),
+			surname: generateRandomString(),
+			id: generateRandomString(),
+		};
+
+		test('should fail if the email already exists', async () => {
```

Comment:
I think all these errors needs to be a query string on the redirected URI

otherwise it probably will be difficult for front end to manage the error handling